src/libmongoc/src/mongoc/mongoc-client-session.c

```diff
@@ -187,6 +218,11 @@ txn_finish (mongoc_client_session_t *session,
    bson_destroy (&reply_local);
    bson_destroy (&cmd);
    bson_destroy (&opts);
+
+   if (retry_wc) {
```

Comment:
Let's move this above line 189 so that, whether you've executed GOTO (retry) or not, you can never leak retry_wc. So:
```
mongoc_write_concern_destroy (retry_wc);
retry_wc = NULL;
if (is_retryable && error_type == MONGOC_WRITE_ERR_RETRY) {
...
```

Reply:
The conditional was based on prior art, as I found null-checking conditionals around most other `mongoc_write_concern_destroy()` calls elsewhere in the driver (e.g. `mongoc-bulk-operation.c`, `mongoc-client.c`, `mongoc-collection.c`). I assumed this was done to save a functional call, but let me know if you still want this removed.

I opted not to remove this above the GOTO for the internal retry attempt, as my most recent commit refactored things a bit to avoid redundantly recreating the WC. For instance, if `txn_commit` is being called for an explicit retry attempt and _that_ first attempt fails with a retryable error we don't need to recreate the WC for the internal retry attempt.